File: nine-language-version/lessons/03-control-flow/control_flow.c
/*
 * Lesson 3: Control Flow in C
 * Demonstrates conditionals, loops, and boolean logic
 */

#include <stdio.h>
#include <stdbool.h>

int main() {
    printf("=== C Control Flow ===\n\n");

    // 1. Basic conditionals
    printf("1. Basic Conditionals:\n");
    int age = 20;
    if (age >= 18) {
        printf("  Age %d: Adult\n", age);
    } else if (age >= 13) {
        printf("  Age %d: Teenager\n", age);
    } else {
        printf("  Age %d: Child\n", age);
    }

    // Ternary expression
    char* status = (age >= 18) ? "Adult" : "Minor";
    printf("  Status: %s\n", status);

    // 2. For loops
    printf("\n2. For Loops:\n");
    printf("  Count to 5:\n   ");
    for (int i = 0; i < 5; i++) {
        printf(" %d", i);
    }
    printf("\n");

    printf("  Iterate array:\n");
    char* fruits[] = {"apple", "banana", "cherry"};
    int num_fruits = 3;
    for (int i = 0; i < num_fruits; i++) {
        printf("    %s\n", fruits[i]);
    }

    // 3. While loop
    printf("\n3. While Loop:\n");
    int count = 0;
    while (count < 3) {
        printf("  Count: %d\n", count);
        count++;
    }

    // 4. Boolean logic
    printf("\n4. Boolean Logic:\n");
    int x = 5, y = 10;
    printf("  x=%d, y=%d\n", x, y);
    printf("  x > 3 && y < 20: %d\n", x > 3 && y < 20);
    printf("  x > 10 || y > 5: %d\n", x > 10 || y > 5);
    printf("  !(x == y): %d\n", !(x == y));

    printf("\n  Truthiness (0 = false, non-zero = true):\n");
    int test_values[] = {0, 1, -1, 42};
    for (int i = 0; i < 4; i++) {
        int val = test_values[i];
        char* truthy = val ? "truthy" : "falsy";
        printf("    %d: %s\n", val, truthy);
    }

    // 5. FizzBuzz
    printf("\n5. FizzBuzz (1-20):\n ");
    for (int i = 1; i <= 20; i++) {
        if (i % 15 == 0) {
            printf(" FizzBuzz");
        } else if (i % 3 == 0) {
            printf(" Fizz");
        } else if (i % 5 == 0) {
            printf(" Buzz");
        } else {
            printf(" %d", i);
        }
    }
    printf("\n");

    // 6. Switch statement
    printf("\n6. Switch Statement:\n");
    int day = 3; // Wednesday
    char* day_type;
    switch (day) {
        case 0:
        case 6:
            day_type = "Weekend";
            break;
        case 1:
        case 2:
        case 3:
        case 4:
        case 5:
            day_type = "Weekday";
            break;
        default:
            day_type = "Invalid day";
    }
    printf("  Day %d: %s\n", day, day_type);

    // 7. Do-while loop
    printf("\n7. Do-While Loop:\n");
    int i = 0;
    do {
        printf("  Iteration: %d\n", i);
        i++;
    } while (i < 3);

    return 0;
}

File: nine-language-version/lessons/03-control-flow/ControlFlow.java
/*
 * Lesson 3: Control Flow in Java
 * Demonstrates conditionals, loops, and boolean logic
 */

public class ControlFlow {
    public static void main(String[] args) {
        System.out.println("=== Java Control Flow ===\n");

        // 1. Basic conditionals
        System.out.println("1. Basic Conditionals:");
        int age = 20;
        if (age >= 18) {
            System.out.println("  Age " + age + ": Adult");
        } else if (age >= 13) {
            System.out.println("  Age " + age + ": Teenager");
        } else {
            System.out.println("  Age " + age + ": Child");
        }

        // Ternary expression
        String status = (age >= 18) ? "Adult" : "Minor";
        System.out.println("  Status: " + status);

        // 2. For loops
        System.out.println("\n2. For Loops:");
        System.out.print("  Count to 5:\n   ");
        for (int i = 0; i < 5; i++) {
            System.out.print(" " + i);
        }
        System.out.println();

        System.out.println("  Iterate array:");
        String[] fruits = {"apple", "banana", "cherry"};
        for (int i = 0; i < fruits.length; i++) {
            System.out.println("    " + fruits[i]);
        }

        System.out.println("  Enhanced for loop:");
        for (String fruit : fruits) {
            System.out.println("    " + fruit);
        }

        // 3. While loop
        System.out.println("\n3. While Loop:");
        int count = 0;
        while (count < 3) {
            System.out.println("  Count: " + count);
            count++;
        }

        // 4. Boolean logic
        System.out.println("\n4. Boolean Logic:");
        int x = 5, y = 10;
        System.out.println("  x=" + x + ", y=" + y);
        System.out.println("  x > 3 && y < 20: " + (x > 3 && y < 20));
        System.out.println("  x > 10 || y > 5: " + (x > 10 || y > 5));
        System.out.println("  !(x == y): " + !(x == y));

        System.out.println("\n  Only boolean type (no truthiness):");
        boolean t = true;
        boolean f = false;
        System.out.println("    true: " + (t ? "truthy" : "falsy"));
        System.out.println("    false: " + (f ? "truthy" : "falsy"));

        // 5. FizzBuzz
        System.out.println("\n5. FizzBuzz (1-20):");
        System.out.print(" ");
        for (int i = 1; i <= 20; i++) {
            if (i % 15 == 0) {
                System.out.print(" FizzBuzz");
            } else if (i % 3 == 0) {
                System.out.print(" Fizz");
            } else if (i % 5 == 0) {
                System.out.print(" Buzz");
            } else {
                System.out.print(" " + i);
            }
        }
        System.out.println();

        // 6. Switch statement
        System.out.println("\n6. Switch Statement:");
        int day = 3; // Wednesday
        String dayType;
        switch (day) {
            case 0:
            case 6:
                dayType = "Weekend";
                break;
            case 1:
            case 2:
            case 3:
            case 4:
            case 5:
                dayType = "Weekday";
                break;
            default:
                dayType = "Invalid day";
        }
        System.out.println("  Day " + day + ": " + dayType);

        // 7. Do-while loop
        System.out.println("\n7. Do-While Loop:");
        int i = 0;
        do {
            System.out.println("  Iteration: " + i);
            i++;
        } while (i < 3);

        // 8. Switch expressions (Java 14+)
        System.out.println("\n8. Switch Expression (Java 14+):");
        try {
            // This uses switch expression syntax
            int score = 85;
            String grade = switch (score / 10) {
                case 10, 9 -> "A";
                case 8 -> "B";
                case 7 -> "C";
                case 6 -> "D";
                default -> "F";
            };
            System.out.println("  Score " + score + ": Grade " + grade);
        } catch (Error e) {
            System.out.println("  (Switch expressions require Java 14+)");
        }
    }
}
